lib/online_order.rb

```diff
@@ -0,0 +1,103 @@
+require 'awesome_print'
+require 'pry'
+require 'csv'
+require_relative 'order'
+require_relative 'customer'
+
+
+module Grocery
+
+# Custom ArgumentError because it's fun, could also be done at the point of raising the ArgumentError using syntax ArgumentError.new("Custom Error Message")
+  class CartError < ArgumentError
+    def initialize(msg="Error: Items cannot be added to this order based on its current status")
+      super
+    end
+  end
+
+  class OnlineOrder < Order
+    attr_reader :customer, :status
+    @@all_online_orders = []
+    @@customers = []
+
+    def initialize(id, products, customer_id, status = :pending)
+      #Use the itialize attributes from the parent class Order
+      super(id, products)
+      @customer = customer_id
+      @status = status.to_sym
+    end # Initialize
+
+    def total
+      unless @products.length == 0
+        return super + 10
+      end
+    end
+
+    def add_product(product_name, product_price)
+      if @status == :paid || @status == :pending
+        super
+      else
+        begin
+          raise CartError
+        rescue
+        end
+      end
+    end
+
+    def self.customers
+      return @@customers
+    end
+
+    def self.all
+      if @@all_online_orders.length == 0
+        self.populate
+      end
+        return @@all_online_orders
+    end
+
+    def self.populate
+      CSV.open('support/online_orders.csv', 'r', headers: true, header_converters: :symbol).each do |row|
+        products = {}
+        id = row[:id].to_i
+        customer_id = row[:customer_id].to_i
+        status = row[:status].to_sym
+        row[:products].split(';').each do |item|
+          name, price = item.split(':')
+          products[name] = price.to_f
+        end
+        @@all_online_orders << index = self.new(id, products, customer_id, status)
+        @@customers << Grocery::Customer.find(customer_id)
+      end
```

Comment:
Line 68 appears to be the cause of your performance problems. Every time you call `OnlineOrder.all`, for each online order in the CSV file, you call `Customer.find`, which involves reading the entire customer CSV file. That means you're opening `customers.csv` a hundred times for every call to `OnlineOrder.all`. Working with a file is one of the slowest things a computer can do, hundreds of thousands or millions of times slower than accessing a value stored in memory, which is why this takes so long.

To find the problem, I `xdescribe`-ed all of your tests, and then re-enabled them group by group. The ones for `OnlineOrder.all` were the only slow ones, which gave me a big hint of where to look.

Interestingly, if you had used the same memoization strategy in `Customer` that you did in `Order` with `@@orders` and `OnlineOrder` with `@@all_online_orders` then this wouldn't have been an issue, since you would only ever open the file once.

If you're interested, [here](https://gist.github.com/jboner/2841832) is some more info on how long it takes a computer to do various things.